### PR TITLE
[POC] feat: validate batch and exclude txs that revert

### LINKED ISSUE #1130: Improve transaction batching
## Bug description

Transactions that modify owners, thresholds or upgrade the Safe should not be able to be batched. The batched transaction will otherwise become stuck.

Edit: we may be able to use the decoding service (for security) for this.

## Steps to reproduce

Queue two transactions that modify the owners and execute them in batch.

## Expected result

Only "standard" transactions should be able to be batched.

## Obtained result

Batched transactions can get stuck.

## Links

- [Slack discussion](https://5afe.slack.com/archives/C03BHKS5K6D/p1668086873310059)
- [Proposed solutions](https://www.notion.so/gnosis-safe/Improving-transaction-batching-88222641f452497589abe10d9dd52701)

### PR DESCRIPTION
**Pull request by iamacook**
## What it solves

Resolves #1130

## How this PR fixes it

The initial batched transaction is validated and, if it fails, each nested transaction is sequentially included and validated inside a new batch. The valid batch is then returned, excluding the transaction that would otherwise fail and those thereafter.

## How to test it

Open a m/n Safe and queue the following transactions in order:

1. Remove owner and set threshold to 1.
2. Set threshold to m.

Transaction 2 is invalid as it attempts to set the threshold higher than that of the Safe is transaction 1 succeeds, which should fail validation.

## Screenshots

image

## Checklist
* [ ] I've tested the branch on mobile 📱
* [ ] I've documented how it affects the analytics (if at all) 📊
* [ ] I've written a unit/e2e test for it (if applicable) 🧑‍💻

### CODE REVIEW

**Comment by katspaugh** [comment]
Awesome idea!
An alternative UI could be: highlight the failing/mutually exclusive sub-transactions in red and add a trash icon to them and suggest to the user to exclude them manually.

**Comment by iamacook** [comment]
I've just pushed a change that validates the transactions syncronously and it is noticeably faster. However, it also validates transactions after invalid one(s) in the proposed batch, resulting in unnecessary RPC calls.

---

I investigated using Tenderly simulations and it is possible using this (unoptimised) snippet:

```ts
const callErrors = simulation.transaction.call_trace.filter((call) => {
  return call.error
})

const firstRevertedTx = allMultiSendTxs.find((tx) => {
  return callErrors.some((error) => tx.data === error.input)
})
```

However, this is not type-correct (we are using unofficial types). (@usame-algan mentioned that Tenderly are currently working on a TypeScript SDK though.)

---

To summarise: the former increases RPC calls, the latter needs more testing (and we'd need to confirm the extra simulation calls don't negatively affect our API limit).

I will leave this as a draft for the time being as we still need to discuss this with design.

File: src/utils/transactions.ts [hunk 1]
```diff
@@ -130,6 +132,40 @@ export const getMultiSendTxs = (
...
+
+  const multiSendTxs = getMultiSendTxs(txs, chain, safeAddress, safeVersion)
+
+  const isValidBatch = await _isValidMultiSend(safe, multiSendTxs)
+
+  if (isValidBatch) {
+    return multiSendTxs
+  }
+
+  const validity = await Promise.all(multiSendTxs.map((_, i) => _isValidMultiSend(safe, multiSendTxs.slice(0, i))))
```
Comment (schmanu) [inline src/utils/transactions.ts:163]: This will always test all possible tx combinations.
I think the most request number efficient implemention would be through binary search:

```ts
  let idx = multiSendTxs.length
  let bestBatch: MetaTransactionData[] = []
  while (idx > 0 && idx <= multiSendTxs.length) {
    const isValid = await _isValidMultiSend(safe, multiSendTxs.slice(0, idx))
    if (isValid) {
      bestBatch = multiSendTxs.slice(0, idx)
      idx += Math.floor(idx / 2)
    } else {
      idx = Math.floor(idx / 2)
    }
  }
  return bestBatch
```

**Review by schmanu — COMMENTED** [review]
Killer feature! 🔪 

After reading your comment I think the tenderly solution would be the cleanest because it has a constant runtime. As both our RPC endpoints and Tenderly have rate limits this should make it quite cost efficient too.
The only issue with using tenderly is that it could be feature-toggled off via our config service. So we would need a fallback for when tenderly is not available. Although the fallback could just be to batch all txs.

**Comment by iamacook** [comment]
> After reading your comment I think the tenderly solution would be the cleanest because it has a constant runtime. As both our RPC endpoints and Tenderly have rate limits this should make it quite cost efficient too.

Valid point. I think this needs to be discussed with the wider team as it would have an impact on our Tenderly rate limit. I think we should look into the usage of batching first though.

> The only issue with using tenderly is that it could be feature-toggled off via our config service. So we would need a fallback for when tenderly is not available. Although the fallback could just be to batch all txs.

I totally agree as well.

I'll come back to this shortly.

**Comment by iamacook** [comment]
@schmanu, I opted to use "our" simulation endpoint as the [Tenderly SDK](https://github.com/Tenderly/tenderly-sdk) requires an access key, which we use a proxy for. I confirmed the designs with @liliiaorlenko as well (see the description for an updated screenshot).

File: src/components/tx/modals/BatchExecuteModal/ReviewBatchExecute.tsx [hunk 2]
```diff
@@ -44,28 +48,65 @@ const ReviewBatchExecute = ({ data, onSubmit }: { data: BatchExecuteData; onSubm
...
+      executionOwner: wallet.address,
+      safe,
+      transactions: allMultiSendTxs,
+    })
+
+    // Use default Tenderly API
+    return getSimulation(simulationPayload, undefined)
+  }, [wallet?.address, safe.address.value, allMultiSendTxs])
+
+  const validMultiSendTxs = useMemo(() => {
```
Comment (schmanu) [inline src/components/tx/modals/BatchExecuteModal/ReviewBatchExecute.tsx:70]: Let's move this logic outside of the component and test it.

File: src/components/tx/modals/BatchExecuteModal/ReviewBatchExecute.tsx [hunk 3]
```diff
@@ -126,7 +181,7 @@ executions from the same Safe."
           </>
         ) : null}
 
-        {multiSendTxs && <TxSimulation canExecute transactions={multiSendTxs} disabled={submitDisabled} />}
+        {allMultiSendTxs && <TxSimulation canExecute transactions={allMultiSendTxs} disabled={submitDisabled} />}
```
Comment (schmanu) [inline src/components/tx/modals/BatchExecuteModal/ReviewBatchExecute.tsx:184]: The simulation should use the `validMultiSendTxs` instead of `allMultiSendTxs`

**Review by schmanu — COMMENTED** [review]
I really like the solution using tenderly.

**Review by schmanu — APPROVED** [review]
🔥

**Review by usame-algan — COMMENTED** [review]
Really cool feature!

I agree with @schmanu suggestion to add a fallback in case Tenderly is disabled i.e. when `simulation` stays undefined. Currently, when I block the simulation request via network panel we don't show the user any info about the batch anymore:

Screenshot 2023-04-14 at 13 27 48

**Comment by iamacook** [comment]
> Really cool feature!
> 
> I agree with @schmanu suggestion to add a fallback in case Tenderly is disabled i.e. when `simulation` stays undefined. Currently, when I block the simulation request via network panel we don't show the user any info about the batch anymore:

I missed that suggestion 🙈 I've adjusted it to now return all transactions if the simulation fails:

image

> As another iteration it would also be nice to have a loading state

I've also added a loader:

loading-batch

File: src/components/tx/modals/BatchExecuteModal/ReviewBatchExecute.tsx [hunk 4]
```diff
@@ -44,28 +66,59 @@ const ReviewBatchExecute = ({ data, onSubmit }: { data: BatchExecuteData; onSubm
...
+    })
+
+    // Use default Tenderly API
+    return getSimulation(simulationPayload, undefined)
+  }, [wallet?.address, safe.address.value, allMultiSendTxs])
+
+  const validMultiSendTxs = useMemo(() => {
+    // Simulation failed, return all txs
+    if (simulationError) {
+      console.log('error')
```
Comment (usame-algan) [inline src/components/tx/modals/BatchExecuteModal/ReviewBatchExecute.tsx:91]:
```suggestion

```

**Review by usame-algan — APPROVED** [review]
Thanks for adding the loader! Looks amazing 💯

**Comment by francovenica** [comment]
Issue:
If the very 1st tx will fail, there is basically no batch, but the send button is still available. Clicking the send button will show a 422 error undefined

The Send button should be disabled if the batch tx list is empty. 
image

**Comment by iamacook** [comment]
After [discussion](https://5afe.slack.com/archives/C03DAGWJCR5/p1681814556445119), we will close this for now and assess the approach again when the modal redesign is implemented.
